nimbus/src/enrollment.rs

```diff
@@ -20,28 +23,300 @@ pub enum EnrolledReason {
+                        &enrollment_id,
+                        &branch,
+                        Some("optout"),
+                        EnrollmentChangeEventType::Disqualification,
+                    ));
+                    Self {
+                        slug: self.slug.clone(),
+                        status: EnrollmentStatus::Disqualified {
+                            reason: DisqualifiedReason::OptOut,
+                            enrollment_id,
+                            branch,
+                        },
+                    }
+                } else {
+                    let mut updated_enrollment = evaluate_enrollment(
+                        nimbus_id,
+                        available_randomization_units,
+                        app_context,
+                        updated_experiment,
+                    )?;
+                    if let EnrollmentStatus::Enrolled { branch, reason, .. } =
+                        updated_enrollment.status
+                    {
+                        // This is a bit dirty, but we want to re-use the enrollment ID we had.
+                        // Q: does it still makes sense to do that on experiments change?
+                        updated_enrollment.status = EnrollmentStatus::Enrolled {
+                            reason,
+                            enrollment_id,
+                            branch,
+                        };
+                        // TODO: do something (e.g. fire an event) on branch change?
+                        updated_enrollment
+                    } else {
+                        log::debug!(
+                            "Existing experiment enrollment '{}' is now disqualified (params change)",
+                            &self.slug
+                        );
+                        enrollment_events.push(EnrollmentChangeEvent::new(
+                            &self.slug,
+                            &enrollment_id,
+                            &branch,
+                            Some("targeting"),
+                            EnrollmentChangeEventType::Disqualification,
+                        ));
+                        Self {
+                            slug: self.slug.clone(),
+                            status: EnrollmentStatus::Disqualified {
+                                reason: DisqualifiedReason::Targeting,
+                                enrollment_id,
+                                branch,
+                            },
+                        }
+                    }
+                }
+            }
+            EnrollmentStatus::Disqualified { .. } => {
+                if !is_user_participating {
+                    log::debug!(
+                        "Disqualified experiment enrollment '{}' has been discarded (global opt-out)",
+                        &self.slug
+                    );
+                    // Discard disqualified enrollments on global experiments opt-out.
+                    return Ok(None);
```

Comment:
Why is this case `None` rather than `EnrollmentStatus::NotEnrolled {reason: NotEnrolledReason::OptOut...}`, like is done in the fn above? Doesn't this mean the experiment will no longer exist in the DB?

Reply:
I implemented what was written in the document: `To Discarded state, when the user entirely disables telemetry in their client.`, but I think I agree with you: if you look closely at the `from_new_experiment` and `on_experiment_updated` (if `self.status == EnrollmentStatus::NotEnrolled`) function bodies, these 2 blocks do functionally the same thing.

Reply:
Note that "when the user entirely disables telemetry" is different from "when the user opts out of experiments" and I'm not sure what API surface we have for the application to indicate this state.